tools/cdi-containerimage-server/main.go

```diff
@@ -0,0 +1,93 @@
+package main
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net"
+	"net/http"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/pkg/errors"
+)
+
+func printFiles(dir string) error {
+	return filepath.Walk(dir, func(path string, info os.FileInfo, err error) error {
+		fmt.Println(path)
+		return nil
+	})
+}
+
+func renameImageFile(dir, newName string) error {
+	entries, err := ioutil.ReadDir(dir)
+	if err != nil {
+		return err
+	}
+
+	if len(entries) != 1 || entries[0].IsDir() {
+		return errors.Errorf("Invalid container image")
+	}
+
+	src := filepath.Join(dir, entries[0].Name())
+	target := filepath.Join(dir, newName)
+
+	if err := os.Rename(src, target); err != nil {
+		return err
+	}
+	return nil
+}
+
+func main() {
+	port := flag.Int("p", 8100, "server port")
+	directory := flag.String("image-dir", ".", "directory to serve")
+	readyFile := flag.String("ready-file", "/shared/ready", "file to create when ready for connections")
+	doneFile := flag.String("done-file", "/shared/done", "file created when the client is done")
+	imageName := flag.String("image-name", "disk.img", "name of the image to serve up")
+	flag.Parse()
+
+	if err := printFiles(*directory); err != nil {
+		log.Fatalf("Failed walking the directory %s: %v", *directory, err)
+	}
+	if err := renameImageFile(*directory, *imageName); err != nil {
+		log.Fatalf("Failed renaming image file %s, directory %s: %v", *imageName, *directory, err)
+	}
+	server := &http.Server{
+		Handler: http.FileServer(http.Dir(*directory)),
+	}
+	addr := fmt.Sprintf("localhost:%d", *port)
+	listener, err := net.Listen("tcp", addr)
+	if err != nil {
+		log.Fatalf("Failed listening on %s err: %v", addr, err)
+	}
+
+	f, err := os.OpenFile(*readyFile, os.O_CREATE|os.O_EXCL, 0666)
```

Comment:
maybe shouldn't have `O_EXCL` here in case of crash/restart.  I think some thought has to go into this synchronization and crash/restart scenarios.  Perhaps the easiest thing to do would be to not restart containers and have the import delete/recreate the pod on failure?

Reply:
Do you really want to move the delete/recreate to pod level by the import-controller, so it adds a specific recovery path to the cri import? I think our server and importer containers are simple enough to do the recovery in the container level. "ready" file removal is deferred by the server, so I see no reason to remove the O_EXCL.

Reply:
I just think we should make sure that we can tolerate container crashes.  If you can make it work without help from the controller, then that's great.  My point was that it may be easier to do that in one place.  

The reason to remove `O_EXCL` is that it is an error if the file already exists and that seems possible after a crash/restart.

Reply:
Ah now I see the defer removal

Reply:
So we can trust the defer removal that when server container crashes the "ready" file is removed, and we can keep the `O_EXCL`. What bothers me more are the loops waiting for the ready/done files. I guess we should error after some time? for the `ready` it's easy (say 10sec?), but for the `done` it depends on the file size.